### PR TITLE
Fixed: avability to create duplicate of an existing category

### LINKED ISSUE #163: An admin can create duplicate of an existing category (if non-printable spaces are used)
**Date Time**
11-17-2020 15:30

**Priority**
Medium

**Description**
An admin can create a duplicate of an existing category in the case of using non-printable spaces in its name.

**Pre-conditions**
Login into EventExpress as an admin 

**Steps to Reproduce**
1. Select Categories
2. Add a new category (by pressing Plus on the top of the screen)
3. Type Summer and then add space (press space bar) before Summer.
4. New Category is created which visually looks the same as already existing (both at the front-end and in the database). 

**Expected result**
The system should check for similarity and suggest if a similar event is already created. 

**Actual result**
An event visually looking similar to an existing one is created. 

**Environment:**
Windows 10 Home OS Build 18363.1139
Google Chrome Version 86.0.4240.111 (Official Build) (64-bit)

**Additional context**

### LINKED ISSUE #328: An invalid notification appears on the 'Categories' page when creating a new category.
**Date Time**
12-13-2020 20:38

**Priority**
Medium

**Environment:**
 OS: Fedora 33 Kernel: 5.9.13-200.fc33.x86_64 x86_64 bits: 64 Desktop: GNOME 3.38.2
 Browser: Google Chrome Version 87.0.4280.88 (Official Build) (64-bit)
 Reproducible: always
 Build found: 20201211.6
 
**Pre-conditions**
1. Move to https://eventsexpress-test.azurewebsites.net/
2. Authorize as an admin

**Steps to Reproduce**
1. Go to the 'Categories' page.
2. Click on a blue “plus-circle" icon on the top of the page with categories.
3. Enter more than 20 characters in the input field labeled "Name".
4. Click on the 'apply' icon

**Test data**
1. 'abracadabraabracadabra'

**Expected result**
The error message appeared: "Name length exceeded the recommended length of 20 characters!" is placed below the “Name” input field and had the red color.

**Actual result**
The invalid error message appeared: `"{"type":"https://tools.ietf.org/html/rfc7231#section-6.5.1","title":"One or more validation errors occurred.","status":400,"traceId":"|35c81d50-4eb26cdb92266f16.","errors":{"Name":["Name length exceeded the recommended length of 20 characters!"]}}"` is placed under the input field and had the red color.

**User story and test case links** 

User story: [#174 As an admin, I want to change the existing category name to update the data.](https://github.com/EventsExpress/EventsExpress/issues/174)
Test case: [#269 Verify that the user as an admin role is able to add the new category name to the list of existing categories.](https://github.com/EventsExpress/EventsExpress/issues/269)

**Labels to be added**
Priority: Medium
Severity: Minor
Type: UI

### LINKED ISSUE #338: Incorrect error message on the "Categories" page while entering short Categories name.
**Date Time**
12-15-2020 12:54

**Priority**
 Low

**Environment:**
 OS: [Windows 10]
Browser [Chrome Version 86.0.4240.198 (Official Build) (64-bit) ]
Reproducible: always
 Build found: 20201214.5
 
**Pre-conditions**
1. Move to https://eventsexpress-test.azurewebsites.net/
2. Authorize as an admin

**Steps to Reproduce**
1. Go to the 'Categories' page.
2. Click on a blue “plus-circle" icon on the top of the page with categories.
3. Enter less than 2 characters in the input field labeled "Name".
4. Click on the 'apply' icon.

**Test data**
'A'

**Expected result**
The error message appeared: "Category confirmation is too short (minimum is 2 characters)" is placed below the “Name” input field and had the red color.

**Actual result**
The incorrect error message appeared: {"type":"https://tools.ietf.org/html/rfc7231#section-6.5.1","title":"One or more validation errors occurred.","status":400,"traceId":"|373dfa09-4844d12e189d4c13.","errors":{"Name":["Name length exceeded the recommended length of 20 characters!"]}}  is placed under the input field and had the red color.

User story and test case links

User story: #174 As an admin, I want to change the existing category name to update the data.
Test case: #269 Verify that the user as an admin role is able to add the new category name to the list of existing categories.

Labels to be added
"Bug"
Priority: Low
Severity: Minor
Type: UI

**Screenshots**

### LINKED ISSUE #339: Incorrect error message on the "Categories" page while entering the same Categories name.
**Date Time**
12-15-2020 13:24

**Priority**
Low

**Environment:**
 OS: [Windows 10]
 Browser [Chrome Version 86.0.4240.198 (Official Build) (64-bit) ]
 Reproducible: always
 Build found: 20201214.5
 
**Pre-conditions**
1. Move to https://eventsexpress-test.azurewebsites.net/
2. Authorize as an admin

**Steps to Reproduce**
1. Go to the 'Categories' page.
2. Click on a blue “plus-circle" icon on the top of the page with categories.
3. Enter same Categories name in the input field labeled "Name".
4. Click on the 'apply' icon.

**Test data**
'golfgolf'

**Expected result**
The error message appeared: "The same category already exist in database" is placed below the “Name” input field and had the red color.

**Actual result**
The incorrect error message appeared: {"":{"rawValue":null,"attemptedValue":null,"errors":[{"exception":null,"errorMessage":"Unhandled exception occurred. Please try again. If this error persists - contact system administrator."}],"validationState":1,"isContainerNode":false,"children":null}} is placed under the input field and had the red color.

User story and test case links

User story: #174 As an admin, I want to change the existing category name to update the data.
Test case: #269 Verify that the user as an admin role is able to add the new category name to the list of existing categories.

Labels to be added
"Bug"
Priority: Low
Severity: Minor
Type: UI

**Screenshots**

Golf

### LINKED ISSUE #340: The error message remains on the Category editbox after canceling invalid Name and clicking on a blue “plus-circle" icon .
**Date Time**
12-15-2020 20:10

**Priority**
 Low

**Environment:**
OS: [Windows 10]
Browser [Chrome Version 86.0.4240.198 (Official Build) (64-bit) ]
Reproducible: always
Build found: 20201214.5
 
**Pre-conditions**
1. Move to https://eventsexpress-test.azurewebsites.net/
2. Authorize as an admin

**Steps to Reproduce**
1. Go to the 'Categories' page.
2. Click on a blue “plus-circle" icon on the top of the page with categories.
3. Enter same Categories name in the input field labeled "Name".
4. Click on the "Create" icon.
5. Click on the “Cancel” icon at the end of the row.
6. Click on a blue “plus-circle" icon on the top of the page with categories.

**Test data**
'golfgolf'

**Expected result**
The input field with label “Name” becomes enabled and is empty by default.

**Actual result**
Error message appeared: {"":{"rawValue":null,"attemptedValue":null,"errors":[{"exception":null,"errorMessage":"Unhandled exception occurred. Please try again. If this error persists - contact system administrator."}],"validationState":1,"isContainerNode":false,"children":null}} is placed under the input field and had the red color.

User story and test case links:
User story: #174 As an admin, I want to change the existing category name to update the data.
Test case: #269 Verify that the user as an admin role is able to add the new category name to the list of existing categories.

Labels to be added
"Bug"
Priority: Low
Severity: Minor
Type: UI

**Screenshots**
Golf
Golf1

### PR DESCRIPTION
**Pull request by Anastasiagumenuk9**
dev
## GitHub Project

* [Main GitHub Project ticket](https://github.com/EventsExpress/EventsExpress/projects/2)

## Code reviewers

- [ ] @iuriikhrystiuk 

### Second Level Review

- [x] @sand0 

## Summary of issue

An admin can create duplicate of an existing category (if non-printable spaces are used)

## Summary of change

- Implemented TrimModelBinder to trim strings from view models
- Edited method for displaying errors
- Move actions for category to a separate folder
- Move reducers for category to a separate folder

## Testing approach

ToDo

## CHECK LIST
- [x]  Сode coverage >=6.98%
- [x]  PR is reviewed manually again (to make sure you have 100% ready code)
- [x]  I've checked new feature as logged in and logged out user if needed
- [x]  PR meets all conventions

### CODE REVIEW

File: EventsExpress/ClientApp/src/services/CategoryService.js [hunk 1]
```diff
@@ -10,7 +10,16 @@ export default class CategoryService {
     }
 
     setCategory = async (data) => {
-        const res = await baseService.setResource('category/edit', {
+        const res = await baseService.setResource(`category/create`, {
+            Name: data.Name
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/services/CategoryService.js]: Consider using `camelCaseNameResolver` when binding.

File: EventsExpress/Startup.cs [hunk 2]
```diff
@@ -149,6 +147,11 @@ public void ConfigureServices(IServiceCollection services)
             services.AddControllers();
             services.AddHttpClient();
 
+            services.Configure<IISServerOptions>(options =>
+            {
+                options.AllowSynchronousIO = true;
```
Comment (iuriikhrystiuk) [inline EventsExpress/Startup.cs]: Why is this required?

File: EventsExpress/EventsExpress.csproj [hunk 3]
```diff
@@ -26,6 +26,7 @@
     <PackageReference Include="MediatR" Version="7.0.0" />
     <PackageReference Include="MediatR.Extensions.Microsoft.DependencyInjection" Version="7.0.0" />
     <PackageReference Include="MicroElements.Swashbuckle.FluentValidation" Version="3.1.0" />
+    <PackageReference Include="Microsoft.AspNet.Mvc" Version="5.2.7" />
```
Comment (sand0) [inline EventsExpress/EventsExpress.csproj]: Is this required?

File: EventsExpress.Core/Services/CategoryService.cs [hunk 4]
```diff
@@ -41,7 +41,7 @@ public IEnumerable<CategoryDTO> GetAllCategories()
 
         public async Task Create(string title)
         {
-            if (_context.Categories.Any(c => c.Name == title))
+            if (ExistsByName(title))
```
Comment (sand0) [inline EventsExpress.Core/Services/CategoryService.cs]: This logic has been added to the validation layer, so can be removed from here

File: EventsExpress.Core/Services/CategoryService.cs [hunk 5]
```diff
@@ -58,7 +58,7 @@ public async Task Edit(CategoryDTO category)
                 throw new EventsExpressException("Not found");
             }
 
-            if (_context.Categories.Any(c => c.Name == category.Name))
+            if (ExistsByName(category.Name))
```
Comment (sand0) [inline EventsExpress.Core/Services/CategoryService.cs]: This logic has been added to the validation layer, so can be removed from here

**Review by sand0 — CHANGES_REQUESTED** [review]

File: EventsExpress/ClientApp/src/services/CategoryService.js [hunk 6]
```diff
@@ -10,17 +10,26 @@ export default class CategoryService {
...
+            name: data.Name
         });
         return !res.ok
             ? { error: await res.text() }
             : res;
     }
 
     setCategoryDelete = async (data) => {
-        const res = await baseService.setResource(`category/delete/${data}`);
+        const res = await baseService.setResource('category/delete/${data}');
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/services/CategoryService.js]: Are we sure we need to change string interpolation into standard string?

File: EventsExpress/ClientApp/src/services/CategoryService.js [hunk 7]
```diff
@@ -10,17 +10,26 @@ export default class CategoryService {
     }
 
     setCategory = async (data) => {
+        const res = await baseService.setResource('category/create', {
+            name: data.Name
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/services/CategoryService.js]: Please consider also changing the naming to camelcase where it is coming from.

**Review by sand0 — APPROVED** [review]

**Pull request merged**